runtime/compiler/aarch64/runtime/PicBuilder.spp

```diff
@@ -47,21 +49,171 @@
+
+	.set	J9TR_USCSnippet_JVMPIEnabled,	0x00800000
+	.set	J9TR_USCSnippet_HelperOffset,	0xFF000000
+	.set	J9TR_USCSnippet_HelperOffsetShift,	24
+
+// Static glue target table is laid out as:
+//
+// 00: 5 helpers
+// 40: 5 sync helpers
+
+	.set	J9TR_SyncJ2ITableOffset,	40
+
 	.text
 	.align 2
 
+// in:     x0  = address of branch instruction
+//         x1  = target of branch
+//         x2  = helper Index
+//         x30 = return address
+//
+// trash:	x3, x4
+
+L_refreshHelper:
+	sub	x3, x1, x0			// distance = (target - callSite)
+	sbfx	x4, x3, #0, #28
+	cmp	x3, x4
+	bne	L_outOfRange			// distance is out of +/-128MB range
+
+	ubfx	x1, x3, #2, #26			// distance >> 2, masking out sign bits
+	mov	w2, #0x14000000			// bit pattern of unconditional branch instruction
+	orr	w2, w2, w1			// embed distance
+	str	w2, [x0]			// rewrite branch instruction
+	mov	x1, #4				// 1 instruction to flush
+	b	flushICache
+
+L_outOfRange:
+	hlt	#0	// Not implemented yet -- Use mcc_lookupHelperTrampoline_unwrapper
+
+// in:     x3  = address of resolve helper function
+//         x30 = snippet
+// out:
+// trash:	x8, x27, x28
+
+L_mergedUnresolvedSpecialStaticGlue:
+	mov	x27, x30					// save snippet address
+	ldr	x0, [x27, #J9TR_SCSnippet_returnAddress]	// Fetch code cache EIP
+	ldr	x1, [x27, #J9TR_USCSnippet_CP]			// get CP
+	ldr	w28, [x27, #J9TR_USCSnippet_CPIndex]		// get CP index & flags (4 bytes)
+	and	x2, x28, #~(J9TR_USCSnippet_HelperOffset)	// remove helper offset from CP index
+	and	x2, x2, #~(J9TR_USCSnippet_JVMPIEnabled)	// remove JVMPI bit from CP index
+	blr	x3						// call resolve helper
+	str	x0, [x27, #J9TR_SCSnippet_method]		// update snippet with resolved method
+	and	x0, x0, #(~clinit_bit)				// clear the clinit bit in the returned address
+	mov	x2, x0						// save method (x0 trashed by following call)
+	bl	jitMethodIsNative				// is the method native?
+	cbz	x0, L_notNative
+	ldr	x1, const_nativeStaticHelper			// if so, use nativeStaticHelper
+	mov	x2, #TR_ARM64nativeStaticHelper
+	b	L_gotHelper					// and skip to writing the address into the instruction
+L_notNative:
+	ldr	x3, const_staticGlueTable			// get helper table address
+	lsr	x1, x28, #J9TR_USCSnippet_HelperOffsetShift	// get helper offset
+	mov	x0, x2						// recover method
+	bl	jitMethodIsSync					// is method synchronized?
+	lsr	x2, x28, #(J9TR_USCSnippet_HelperOffsetShift+1)	// save helper offset for refreshHelper
+	cbz	x0, L_notSync
+	add	x1, x1, #J9TR_SyncJ2ITableOffset		// if so, adjust helper offset
+	add	x2, x2, #1
+L_notSync:
+	add	x2, x2, #TR_ARM64interpreterVoidStaticGlue
+	ldr	x1, [x3, x1]					// fetch static glue helper from table
+L_gotHelper:
+	ldr	x0, [x27, #J9TR_SCSnippet_returnAddress]	// Fetch code cache EIP
+	ldr	x3, [x27, #J9TR_SCSnippet_method]
+	tbnz	x3, #1, L_USSGclinitCase			// branch if the LSB (the "clinit" bit) was set in the resolved address
+	ldr	x28, [x27, #J9TR_USCSnippet_CP]			// get CP
+	ldr	x8, [x27, #J9TR_USCSnippet_CPIndex]		// get CP index & flags
```

Comment:
The cpIndex is already in x28 from above and I don't believe any path has clobbered it yet.  Why not use that instead of loading?  You'll have to find a different register for the cpAddr since you just loaded it into x28.